source/common/http/filter_manager.cc

```diff
@@ -892,6 +892,13 @@ FilterManager::commonDecodePrefix(ActiveStreamDecoderFilter* filter,
 }
 
 void DownstreamFilterManager::onLocalReply(StreamFilterBase::LocalReplyData& data) {
+  if (Runtime::runtimeFeatureEnabled(
+          "envoy.reloadable_features.on_local_reply_createfilterchain")) {
+    // To ensure we have filters over which we can iterate and call onLocalReply.
+    // If the filter chain already exists this will be a no-op.
+    createFilterChain();
```

Comment:
I saw this function is already called below in  local reply code path:

https://github.com/envoyproxy/envoy/blob/bd40760d62ab50a54823271c449670c2e787b5c8/source/common/http/filter_manager.cc#L1032

and here

https://github.com/envoyproxy/envoy/blob/bd40760d62ab50a54823271c449670c2e787b5c8/source/common/http/filter_manager.cc#L982C1-L982C1

Is there a need to add it here?

Reply:
The call to `onLocalReply` occurs before each of those linked functions are called ([here](https://github.com/envoyproxy/envoy/blob/bd40760d62ab50a54823271c449670c2e787b5c8/source/common/http/filter_manager.cc#L924)). So at the time that `onLocalReply` is executed, there is not yet a filter chain constructed.

Reply:
I see. So you want to create filter chain here, so onLocalReply() can be called by ext_proc filter.

Reply:
That is correct, yes. But it seems that some mobile functionality and/or tests were reliant on this bug

Reply:
It looks like Envoy mobile and some other filters already had hooks to onLocalReply(), for example:
https://github.com/envoyproxy/envoy/blob/bd40760d62ab50a54823271c449670c2e787b5c8/mobile/library/common/extensions/filters/http/local_error/filter.cc#L12

It appears they can be called without adding createFilterChain() in onLocalReply() in filter_manager.cc.

Reply:
This is because `sendLocalReply` ([link](https://github.com/envoyproxy/envoy/blob/main/source/common/http/filter_manager.cc#L906)) can be (and is) called from codepaths which have already constructed filter chains. In the case of a protocol error, however, it is invoked directly from the codec, and there is no filter chain yet. This happens when the `ConnectionManagerImpl::onData` ([link](https://github.com/envoyproxy/envoy/blob/main/source/common/http/conn_manager_impl.cc#L465)) calls the codec's `dispatch` ([link](https://github.com/envoyproxy/envoy/blob/main/source/common/http/http1/codec_impl.cc#L1288)), into `dispatchSlice` ([link](https://github.com/envoyproxy/envoy/blob/main/source/common/http/http1/codec_impl.cc#L725)), which calls `sendProtocolError` ([link](https://github.com/envoyproxy/envoy/blob/main/source/common/http/http1/codec_impl.cc#L1373)), which directly invokes `sendLocalReply` ([link](https://github.com/envoyproxy/envoy/blob/main/source/common/http/filter_manager.cc#L455)) on the decoder, which in turn calls its parent `sendLocalReply` ([link](https://github.com/envoyproxy/envoy/blob/main/source/common/http/filter_manager.cc#L906)) where this ordering problem appears.